NJT OpInfo tests v2

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140161
* __->__ #138370
* #140160

This PR updates OpInfo-based tests for NJTs:
* Adds extensive coverage across non-contiguous NJTs (both non-contiguous transposed and non-contiguous with holes)
    * The `_sample_njts()` helper that `sample_input_func`s utilize now produces non-contig NJTs as well
* Utilizes a `SampleInput`-based xfail system for granular classification of bugs. For example, it's possible to indicate that a class of ops is expected to fail only on non-contig with holes NJT inputs.
    * I decided on adding `SampleInput`s and utilizing this system over using test parametrization for two reasons:
        * Test perf - adding `SampleInput`s is faster than generating entire new tests
        * Avoiding the possibility of `sample_input_func`s not respecting the non-contig test parameter - this would result in silently incorrect passing of these tests. Keeping the responsibility for `SampleInput` generation firmly within each `OpInfo`'s `sample_input_func` means weirdness like this isn't possible
* Improves `SampleInput` naming for a bunch of `sample_input_func`s. This makes it easier to xfail them as needed. For example, binary / unary / other ops now use the new `_describe_njt()` helper to get a string repr that uniquely defines the type of NJT being passed to the op
* Adds appropriate `XFailRule`s to get tests passing for forward / backward / forward compile / backward compile. In general, each xfail corresponds to some bug that needs to be fixed

```python
# Represents a rule indicating how to xfail a particular test. It allows granularity
# at the device, dtype, op, and individual sample levels. This flexibility allows entire
# bugs to be represented by a single rule, even if this corresponds with multiple conceptual
# test cases across multiple ops.
@dataclass
class XFailRule:
    # expected error type
    error_type: TypeVar = Exception
    # expected error message
    error_msg: str = ".*"
    # function to indicate whether the rule applies; return True if so
    match_fn: Callable[[torch.device, torch.dtype, OpInfo, SampleInput], bool] = None
    # optional name for identifying the rule
    name: str = ""

    def match(self, device, dtype, op, sample) -> bool:
        return self.match_fn(device, dtype, op, sample)
```

Example:
```python
    # Bug when broadcasting a binary op with non-contiguous with holes NJT + dense
    # tensor with 1 in ragged dim.
    XFailRule(
        error_type=RuntimeError,
        error_msg="cannot call binary pointwise function .* with inputs of shapes",
        match_fn=lambda device, dtype, op, sample: (
            isinstance(op, BinaryUfuncInfo)
            and "noncontig_holes" in sample.name
            and "broadcasting 1 over ragged" in sample.name
        ),
        name="binary_noncontig_holes_broadcasting_1_over_ragged",
    ),
```

jbschlosser committed:
NJT OpInfo tests v2

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"




[ghstack-poisoned]

soulitzer reviewed (approved):
This is great! Awesome work!

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        name="no_linear_noncontig_holes_support",
+    ),
+    # Some kinda reduction bug that needs to be fixed!
+    XFailRule(
```
It feels like this top most section of xfails is like the "expected" section, is it worth moving this bug below somewhere (or does it fit better here for other reasons?)

soulitzer replied:
Don't want this type of stuff to block this PR, but re https://github.com/pytorch/pytorch/pull/138370/files#r1834994353, more of just curious of what the plan of organization is.

jbschlosser replied:
(see discussion below on xfails, resolving this one)

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
-    # === BEGIN BUG SECTION ===
-    # Returns a tuple of Tensors so it doesn't work with NJT's unary pointwise logic
-    "frexp",
+FORWARD_FAILURES = [
```
Maybe worth adding a short comment on how you tend to think about organizing the xfails, or like a table of contents? :P

soulitzer replied:
Reading some more, table of contents seems unrealistic lol

jbschlosser replied:
> Maybe worth adding a short comment on how you tend to think about organizing the xfails

I think this is a good point and something I should do here. You've managed to uncover a sorta tricky detail here - rules are attempted to be matched in list order, so higher-up xfails have higher precedent, so order actually matters! I'm not sure this is actually that great of a property, as it can be hard to maintain these in the right order. So some possibilities here:
1. Keep as-is and describe the ordering property, etc.
2. Always try to match all rules and fail if multiple rules match to ensure rules are mutually exclusive (this can make them slightly harder to write).
3. ???

soulitzer replied:
Ah that is tricky..
(2) is actually a great property, being able to see the error message and being able to reliably map it to a single bug means our error messages are good, but I guess the issue is here is that you might run into some generic compiler error message that you don't have control over and is always going to be ambiguous.

soulitzer replied:
Hmm one way to do it incrementally could be to have a `mutually_exclusive` property that is by default True, but False for the ones that are ambiguous.

jbschlosser replied:
> (2) is actually a great property, being able to see the error message and being able to reliably map it to a single bug means our error messages are good and descriptive, but I guess the issue is here is that you might run into some generic compiler error message that you don't have control over and is always going to be ambiguous.

I think to clarify some: we should always be able to write rules that are not ambiguous wrt how they match. That is, matching can be completely mutually exclusive so no single sample matches to multiple rules, but this can make the rules require detailed checks to ensure this property.

Whether or not error messages are clear is kinda orthogonal to rule matching, although also a good point that it's a nice property to be able to search the XFail db for an encountered error message.

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+    # expected: masked ops don't support jagged layout
+    XFailRule(
+        error_type=ValueError,
+        error_msg="expects strided",
```
Interesting that this doesn't give us "not_implemented". Not really a big deal, but would've been nice to have a mega xfail for the not implemented stuff lol

jbschlosser replied:
yeah, super annoying :(

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+                    "clamp_min",
+                    "float_power",
+                    "pow",
+                    "sinc",
```
this doesn't look like a binary op :P should we move it out into a separate XFailRule?

jbschlosser replied:
haha good point - I think I'll just rename the rule to be more accurate

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        and "with_seqlen_cache" in sample.name,
+        name="atanh_unimplemented_fill",
+    ),
+    # autodiff on complex dtype is not supported
```
Expected:

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        error_type=RuntimeError,
+        error_msg="not supported for nested tensors with holes",
+        match_fn=lambda device, dtype, op, sample: (
+            op.full_name in {"bmm", "matmul"}
```
I wonder if its better to have `ops={"bmm", "matmul"}` be a field on XFailRule, and call `match_fn` something like `sample_match_fn` instead.

With those two separate I can utilize the "xfail_db" to query xfails based on op and transposition/contiguity separately, e.g. what are all the xfails for this op, what are all the xfails that I can run into for non-contig transposed njts.

jbschlosser replied:
I don't necessarily disagree with a hierarchy of xfails, but note that this specific approach does limit the flexibility some. For example, I regularly do things like check `isinstance(op, BinaryUfuncInfo)` which wouldn't be possible if we just had a list of op names.

To keep the existing flexibility: we could have both `op_fn` and `sample_fn`, where the former only needs to be called once per test and the latter needs to be called once per sample. This could help test runtime a tad (TBD how much).

> what are all the xfails that I can run into for non-contig transposed njts.

the problem with this is that most non-contig related xfails only makes sense within the context of a certain class of ops, so I'm not sure we can detangle these entirely.

jbschlosser committed:
Update on "NJT OpInfo tests v2"


This PR updates OpInfo-based tests for NJTs:
* Adds extensive coverage across non-contiguous NJTs (both non-contiguous transposed and non-contiguous with holes)
    * The `_sample_njts()` helper that `sample_input_func`s utilize now produces non-contig NJTs as well
* Utilizes a `SampleInput`-based xfail system for granular classification of bugs. For example, it's possible to indicate that a class of ops is expected to fail only on non-contig with holes NJT inputs.
    * I decided on adding `SampleInput`s and utilizing this system over using test parametrization for two reasons:
        * Test perf - adding `SampleInput`s is faster than generating entire new tests
        * Avoiding the possibility of `sample_input_func`s not respecting the non-contig test parameter - this would result in silently incorrect passing of these tests. Keeping the responsibility for `SampleInput` generation firmly within each `OpInfo`'s `sample_input_func` means weirdness like this isn't possible
* Improves `SampleInput` naming for a bunch of `sample_input_func`s. This makes it easier to xfail them as needed. For example, binary / unary / other ops now use the new `_describe_njt()` helper to get a string repr that uniquely defines the type of NJT being passed to the op
* Adds appropriate `XFailRule`s to get tests passing for forward / backward / forward compile / backward compile. In general, each xfail corresponds to some bug that needs to be fixed

```python
# Represents a rule indicating how to xfail a particular test. It allows granularity
# at the device, dtype, op, and individual sample levels. This flexibility allows entire
# bugs to be represented by a single rule, even if this corresponds with multiple conceptual
# test cases across multiple ops.
dataclass
class XFailRule:
    # expected error type
    error_type: TypeVar = Exception
    # expected error message
    error_msg: str = ".*"
    # function to indicate whether the rule applies; return True if so
    match_fn: Callable[[torch.device, torch.dtype, OpInfo, SampleInput], bool] = None
    # optional name for identifying the rule
    name: str = ""

    def match(self, device, dtype, op, sample) -> bool:
        return self.match_fn(device, dtype, op, sample)
```

Example:
```python
    # Bug when broadcasting a binary op with non-contiguous with holes NJT + dense
    # tensor with 1 in ragged dim.
    XFailRule(
        error_type=RuntimeError,
        error_msg="cannot call binary pointwise function .* with inputs of shapes",
        match_fn=lambda device, dtype, op, sample: (
            isinstance(op, BinaryUfuncInfo)
            and "noncontig_holes" in sample.name
            and "broadcasting 1 over ragged" in sample.name
        ),
        name="binary_noncontig_holes_broadcasting_1_over_ragged",
    ),
```

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"


This PR updates OpInfo-based tests for NJTs:
* Adds extensive coverage across non-contiguous NJTs (both non-contiguous transposed and non-contiguous with holes)
    * The `_sample_njts()` helper that `sample_input_func`s utilize now produces non-contig NJTs as well
* Utilizes a `SampleInput`-based xfail system for granular classification of bugs. For example, it's possible to indicate that a class of ops is expected to fail only on non-contig with holes NJT inputs.
    * I decided on adding `SampleInput`s and utilizing this system over using test parametrization for two reasons:
        * Test perf - adding `SampleInput`s is faster than generating entire new tests
        * Avoiding the possibility of `sample_input_func`s not respecting the non-contig test parameter - this would result in silently incorrect passing of these tests. Keeping the responsibility for `SampleInput` generation firmly within each `OpInfo`'s `sample_input_func` means weirdness like this isn't possible
* Improves `SampleInput` naming for a bunch of `sample_input_func`s. This makes it easier to xfail them as needed. For example, binary / unary / other ops now use the new `_describe_njt()` helper to get a string repr that uniquely defines the type of NJT being passed to the op
* Adds appropriate `XFailRule`s to get tests passing for forward / backward / forward compile / backward compile. In general, each xfail corresponds to some bug that needs to be fixed

```python
# Represents a rule indicating how to xfail a particular test. It allows granularity
# at the device, dtype, op, and individual sample levels. This flexibility allows entire
# bugs to be represented by a single rule, even if this corresponds with multiple conceptual
# test cases across multiple ops.
dataclass
class XFailRule:
    # expected error type
    error_type: TypeVar = Exception
    # expected error message
    error_msg: str = ".*"
    # function to indicate whether the rule applies; return True if so
    match_fn: Callable[[torch.device, torch.dtype, OpInfo, SampleInput], bool] = None
    # optional name for identifying the rule
    name: str = ""

    def match(self, device, dtype, op, sample) -> bool:
        return self.match_fn(device, dtype, op, sample)
```

Example:
```python
    # Bug when broadcasting a binary op with non-contiguous with holes NJT + dense
    # tensor with 1 in ragged dim.
    XFailRule(
        error_type=RuntimeError,
        error_msg="cannot call binary pointwise function .* with inputs of shapes",
        match_fn=lambda device, dtype, op, sample: (
            isinstance(op, BinaryUfuncInfo)
            and "noncontig_holes" in sample.name
            and "broadcasting 1 over ragged" in sample.name
        ),
        name="binary_noncontig_holes_broadcasting_1_over_ragged",
    ),
```

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT OpInfo tests v2"


This PR updates OpInfo-based tests for NJTs:
* Adds extensive coverage across non-contiguous NJTs (both non-contiguous transposed and non-contiguous with holes)
    * The `_sample_njts()` helper that `sample_input_func`s utilize now produces non-contig NJTs as well
* Utilizes a `SampleInput`-based xfail system for granular classification of bugs. For example, it's possible to indicate that a class of ops is expected to fail only on non-contig with holes NJT inputs.
    * I decided on adding `SampleInput`s and utilizing this system over using test parametrization for two reasons:
        * Test perf - adding `SampleInput`s is faster than generating entire new tests
        * Avoiding the possibility of `sample_input_func`s not respecting the non-contig test parameter - this would result in silently incorrect passing of these tests. Keeping the responsibility for `SampleInput` generation firmly within each `OpInfo`'s `sample_input_func` means weirdness like this isn't possible
* Improves `SampleInput` naming for a bunch of `sample_input_func`s. This makes it easier to xfail them as needed. For example, binary / unary / other ops now use the new `_describe_njt()` helper to get a string repr that uniquely defines the type of NJT being passed to the op
* Adds appropriate `XFailRule`s to get tests passing for forward / backward / forward compile / backward compile. In general, each xfail corresponds to some bug that needs to be fixed

```python
# Represents a rule indicating how to xfail a particular test. It allows granularity
# at the device, dtype, op, and individual sample levels. This flexibility allows entire
# bugs to be represented by a single rule, even if this corresponds with multiple conceptual
# test cases across multiple ops.
dataclass
class XFailRule:
    # expected error type
    error_type: TypeVar = Exception
    # expected error message
    error_msg: str = ".*"
    # function to indicate whether the rule applies; return True if so
    match_fn: Callable[[torch.device, torch.dtype, OpInfo, SampleInput], bool] = None
    # optional name for identifying the rule
    name: str = ""

    def match(self, device, dtype, op, sample) -> bool:
        return self.match_fn(device, dtype, op, sample)
```

Example:
```python
    # Bug when broadcasting a binary op with non-contiguous with holes NJT + dense
    # tensor with 1 in ragged dim.
    XFailRule(
        error_type=RuntimeError,
        error_msg="cannot call binary pointwise function .* with inputs of shapes",
        match_fn=lambda device, dtype, op, sample: (
            isinstance(op, BinaryUfuncInfo)
            and "noncontig_holes" in sample.name
            and "broadcasting 1 over ragged" in sample.name
        ),
        name="binary_noncontig_holes_broadcasting_1_over_ragged",
    ),
```

[ghstack-poisoned]

Skylion007 commented on test/test_nestedtensor.py:
```diff
@@ -7824,131 +7832,558 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+            "nn.functional.softplus",
+            "nn.functional.softshrink",
+            "nn.functional.threshold",
+            "rad2deg",
```
huh, Why is rad2deg in this list, but not deg2rad? o-o

jbschlosser replied:
good question! Looks like it's because `deg2rad` is tagged with `tags: pointwise` and `rad2deg` is not for some reason.. this should probably be fixed.

jbschlosser committed:
Update on "NJT OpInfo tests v2"


This PR updates OpInfo-based tests for NJTs:
* Adds extensive coverage across non-contiguous NJTs (both non-contiguous transposed and non-contiguous with holes)
    * The `_sample_njts()` helper that `sample_input_func`s utilize now produces non-contig NJTs as well
* Utilizes a `SampleInput`-based xfail system for granular classification of bugs. For example, it's possible to indicate that a class of ops is expected to fail only on non-contig with holes NJT inputs.
    * I decided on adding `SampleInput`s and utilizing this system over using test parametrization for two reasons:
        * Test perf - adding `SampleInput`s is faster than generating entire new tests
        * Avoiding the possibility of `sample_input_func`s not respecting the non-contig test parameter - this would result in silently incorrect passing of these tests. Keeping the responsibility for `SampleInput` generation firmly within each `OpInfo`'s `sample_input_func` means weirdness like this isn't possible
* Improves `SampleInput` naming for a bunch of `sample_input_func`s. This makes it easier to xfail them as needed. For example, binary / unary / other ops now use the new `_describe_njt()` helper to get a string repr that uniquely defines the type of NJT being passed to the op
* Adds appropriate `XFailRule`s to get tests passing for forward / backward / forward compile / backward compile. In general, each xfail corresponds to some bug that needs to be fixed

```python
# Represents a rule indicating how to xfail a particular test. It allows granularity
# at the device, dtype, op, and individual sample levels. This flexibility allows entire
# bugs to be represented by a single rule, even if this corresponds with multiple conceptual
# test cases across multiple ops.
dataclass
class XFailRule:
    # expected error type
    error_type: TypeVar = Exception
    # expected error message
    error_msg: str = ".*"
    # function to indicate whether the rule applies; return True if so
    match_fn: Callable[[torch.device, torch.dtype, OpInfo, SampleInput], bool] = None
    # optional name for identifying the rule
    name: str = ""

    def match(self, device, dtype, op, sample) -> bool:
        return self.match_fn(device, dtype, op, sample)
```

Example:
```python
    # Bug when broadcasting a binary op with non-contiguous with holes NJT + dense
    # tensor with 1 in ragged dim.
    XFailRule(
        error_type=RuntimeError,
        error_msg="cannot call binary pointwise function .* with inputs of shapes",
        match_fn=lambda device, dtype, op, sample: (
            isinstance(op, BinaryUfuncInfo)
            and "noncontig_holes" in sample.name
            and "broadcasting 1 over ragged" in sample.name
        ),
        name="binary_noncontig_holes_broadcasting_1_over_ragged",
    ),
```

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
